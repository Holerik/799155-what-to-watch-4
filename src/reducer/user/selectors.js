// selectors.js
import NameSpace from '../name-space/name-space.js';

const NAME_SPACE = NameSpace.USER;

export const getAuthorizationStatus = (state) => {
  return state[NAME_SPACE].authorizationStatus;
};

export const getAuthorInfo = (state) => {
  return {
    avatar: state[NAME_SPACE].avatar,
    name: state[NAME_SPACE].authorName,
  };
};
