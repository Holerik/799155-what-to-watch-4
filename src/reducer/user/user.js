// user.js
import {extend} from '../../utils.js';

const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const ActionType = {
  REQUIRED_AUTHORIZATION: `REQUIRED_AUTHORIZATION`,
  SET_AUTHOR_INFORMATION: `SET_AUTHOR_INFORMATION`,
};

const initilalState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  authorId: -1,
  authorName: ``,
  authorEmail: ``,
  avatar: `img/avatar.jpg`,
};

const ActionCreator = {
  requireAuthorization: (status) => {
    return {
      type: ActionType.REQUIRED_AUTHORIZATION,
      payload: status,
    };
  },
  setAuthorInfo: (info) => {
    return {
      type: ActionType.SET_AUTHOR_INFORMATION,
      payload: info,
    };
  },
};

const reducer = (state = initilalState, action) => {
  switch (action.type) {
    case ActionType.REQUIRED_AUTHORIZATION:
      return extend(state, {
        authorizationStatus: action.payload,
      });
    case ActionType.SET_AUTHOR_INFORMATION:
      return extend(state, {
        authorId: action.payload.id,
        authorName: action.payload.name,
        authorEmail: action.payload.email,
        avatar: action.payload.avatar_url,
      });
  }
  return state;
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
      .then(() => {
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
      });
  },
  login: (authData) => (dispatch, getState, api) => {
    return api.post(`/login`, {
      email: authData.email,
      password: authData.password,
    })
    .then(() => {
      dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
    });
  },
};


export {reducer, Operation, ActionType, ActionCreator, AuthorizationStatus};
