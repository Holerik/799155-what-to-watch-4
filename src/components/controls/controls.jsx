// controls.jsx
import React from 'react';
import {Link} from 'react-router-dom';
import PropTypes from 'prop-types';
import {fullInfo} from '../../reducer/data/data.js';
import {AppRoutes} from '../../const.js';
import {connect} from 'react-redux';
import {getMovie} from '../../reducer/movie/selectors.js';

const Controls = React.memo(function Controls(props) {
  const {favoriteButtonClickHandler, movieInfo} = props;
  return (
    <React.Fragment>
      <button className="btn btn--list movie-card__button"
        type="button" onClick={() => favoriteButtonClickHandler(movieInfo)}
      >
        <svg viewBox="0 0 19 20" width="19" height="20">
          {movieInfo.favorite ? <use xlinkHref="#in-list"></use> : <use xlinkHref="#add"></use>}
        </svg>
        <span>My list</span>
      </button>
      <Link to={`${AppRoutes.ADD_REVIEW}/${movieInfo.id}`}
        className="btn movie-card__button">Add review
      </Link>
    </React.Fragment>
  );
}
);

Controls.propTypes = {
  favoriteButtonClickHandler: PropTypes.func.isRequired,
  movieInfo: PropTypes.exact(fullInfo).isRequired,
};

const mapStateToProps = (state) => ({
  movieInfo: getMovie(state),
});

export {Controls};
export default connect(mapStateToProps)(Controls);
