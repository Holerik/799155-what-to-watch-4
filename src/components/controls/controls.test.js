// controls.test.js
import React from 'react';
import renderer from 'react-test-renderer';
import {Controls} from './controls.jsx';
import {AuthorizationStatus} from '../../reducer/user/user.js';
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";
import history from '../../history.js';
import {Router} from 'react-router-dom';

const film = {
  id: 1,
  title: `Joker`,
  poster: `img/joker.jpg`,
  altPoster: `Joker poster`,
  background: `img/joker-bg.jpg`,
  altBackground: `Gotham City`,
  genre: [`Thriller`],
  year: 2019,
  duration: `2h 2min`,
  age: `18+`,
  src: ``,
  preview: ``,
  rating: {
    score: `4.3`,
    level: `very good`,
    count: 150,
  },
  director: `Todd Phillips`,
  starring: [`Joaquin Phoenix`],
  description: ``,
  review: ``,
  reviews: [0],
  favorite: false,
};

const mockStore = configureStore([]);

describe(`Controls test`, () => {
  it(`<Controls /> should render correctly`, () => {
    const store = mockStore({
      USER: {
        authorizationStatus: AuthorizationStatus.NO_AUTH,
      },
      ERROR: {
        message: ``,
        show: false,
      },
    });

    const tree = renderer
    .create(
        <Provider store={store}>
          <Router history={history}>
            <Controls
              favoriteButtonClickHandler= {() => {}}
              movieInfo={film}
              authorizationStatus={AuthorizationStatus.NO_AUTH}
            />
          </Router>
        </Provider>
        , {
          createNodeMock: () => {
            return {};
          }
        }).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
