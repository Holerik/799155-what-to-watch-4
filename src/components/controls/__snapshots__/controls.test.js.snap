// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls test <Controls /> should render correctly 1`] = `
<div
  className="btn btn--list movie-card__button"
>
  <a
    className="logo__link"
    href="/login"
    onClick={[Function]}
  >
    <span>
      My list
    </span>
  </a>
</div>
`;
